lib/Driver/Driver.cpp

```diff
@@ -339,9 +339,36 @@ static bool populateOutOfDateMap(InputInfoMap &map, StringRef argsHashStr,
     map[inputPair.second] = iter->getValue();
   }
 
-  // If a file was removed, we've lost its dependency info. Rebuild everything.
-  // FIXME: Can we do better?
-  return numInputsFromPrevious != previousInputs.size();
+  if (numInputsFromPrevious == previousInputs.size()) {
+    return false;
+  } else {
+    // If a file was removed, we've lost its dependency info. Rebuild everything.
+    // FIXME: Can we do better?
+    if (ShowIncrementalBuildDecisions) {
+      llvm::StringSet<> inputArgs;
+      for (auto &inputPair : inputs) {
+        inputArgs.insert(inputPair.second->getValue());
+      }
+
+      llvm::StringSet<> missingInputs;
+      for (auto &previousInput : previousInputs) {
+        auto previousInputArg = previousInput.getKey();
+        if(inputArgs.find(previousInputArg) == inputArgs.end()) {
+          missingInputs.insert(previousInputArg);
+        }
+      }
+
+      llvm::outs() << "Incremental compilation has been disabled, because "
+                   << "inputs used in the previous compilation were not used "
+                   << "in the current compilation.\n"
+                   << "\tInputs not included in this compilation:";
+      for (auto &missing : missingInputs) {
+        llvm::outs() << " " << missing.getKey();
```

Comment:
Most of the time these are going to be absolute paths, so I suggest putting them on their own lines.